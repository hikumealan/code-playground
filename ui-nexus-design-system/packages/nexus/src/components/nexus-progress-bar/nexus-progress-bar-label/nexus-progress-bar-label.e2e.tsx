
import { newE2EPage } from '@stencil/core/testing';

describe('nexus-progress-bar-label', () => {
  it('renders', async () => {
    const page = await newE2EPage();

    await page.setContent('<nexus-progress-bar-label></nexus-progress-bar-label>');
    const element = await page.find('nexus-progress-bar-label');
    expect(element).toHaveClass('hydrated');
  });
});

