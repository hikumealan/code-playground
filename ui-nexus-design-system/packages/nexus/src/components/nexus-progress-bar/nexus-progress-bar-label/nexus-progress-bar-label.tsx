
import { Component, h, Host } from '@stencil/core';

@Component({
  tag: 'nexus-progress-bar-label',
  shadow: false,
  styleUrl: 'nexus-progress-bar-label.scss'
})
export class NexusProgressBarLabel {
  render() {
    return <Host class="nexus-progress-bar-label"><slot /></Host>;
  }
}

