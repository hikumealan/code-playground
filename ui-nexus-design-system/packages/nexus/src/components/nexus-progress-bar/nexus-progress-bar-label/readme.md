# nexus-progress-bar-label



<!-- Auto Generated Below -->


----------------------------------------------


