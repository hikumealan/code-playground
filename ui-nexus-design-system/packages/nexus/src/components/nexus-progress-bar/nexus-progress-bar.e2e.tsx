
import { newE2EPage } from '@stencil/core/testing';

describe('nexus-progress-bar', () => {
  it('renders', async () => {
    const page = await newE2EPage();

    await page.setContent('<nexus-progress-bar></nexus-progress-bar>');
    const element = await page.find('nexus-progress-bar');
    expect(element).toHaveClass('hydrated');
  });
});

