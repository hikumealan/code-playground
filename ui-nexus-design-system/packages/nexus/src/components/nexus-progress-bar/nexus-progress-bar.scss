@import '../../styles/variables';

.nexus-progress-bar {
  display: flex;
  flex-direction: column;

  .nexus-progress-bar-meter-container {
    display: inline-block;
    height: 4px;
    width: 100%;
    border-radius: 2px;
    background-color: $grey-200;
    position: relative;
    overflow: hidden;
  }

  .nexus-progress-bar-meter {
    top: 0;
    bottom: 0;
    position: absolute;
    border-radius: 2px;
    background-color: var(--primary-accent);

    &.nexus-progress-bar-meter-error {
      background-color: var(--error);
    }

    &.nexus-progress-bar-meter-indeterminate {
      animation: grow 1.5s linear infinite;

      @keyframes grow {
        0% {
          width: 0;
        }

        100% {
          width: 100%;
        }
      }
    }
  }

  @mixin size($width, $height) {
    height: $height;
    width: $width;
  }

  &.nexus-progress-bar-circle {
    @include size(var(--progress-bar-circle-size), var(--progress-bar-circle-size));

    position: relative;
  }

  .nexus-progress-bar-pie {
    @include size(100%, 100%);


    left: 0;
    position: absolute;
    top: 0;

    .nexus-progress-bar-left-hidden {
      position: absolute;
      width: 100%;
      height: 100%;
      border-radius: 50%;
      border: var(--progress-bar-circle-border-width) solid var(--progress-bar-bg-color);
      clip: rect(0, 41px, 81px, 0);
    }

    &.nexus-progress-bar-hide-right {
      .nexus-progress-bar-right-side {
        display: none;
      }
    }

    &:not(.nexus-progress-bar-hide-right) {
      clip: rect(auto, auto, auto, auto);

      .nexus-progress-bar-right-side {
        transform: rotate(180deg);
      }
    }

    &.nexus-progress-bar-error {
      .nexus-progress-bar-half-circle {
        border-color: var(--error);
      }
    }

    .nexus-progress-bar-half-circle {
      @include size(100%, 100%);

      border: var(--progress-bar-circle-border-width) solid var(--primary-accent);
      border-radius: 50%;
      clip: rect(0, var(--progress-bar-circle-diameter-size), var(--progress-bar-circle-size), 0);
      left: 0;
      position: absolute;
      top: 0;
    }

    .nexus-progress-bar-label {
      height: 100%;
      display: flex;
      justify-content: center;
      align-items: center;
    }
  }

  .nexus-progress-bar-shadow {
    @include size(100%, 100%);

    border: var(--progress-bar-circle-border-width) solid var(--progress-bar-bg-color);
    border-radius: 50%;
  }
}

.nexus-progress-bar-label-span {
  font-size: var(--progress-bar-label-span-font-size);
  color: var(--progress-bar-label-span-color);
}

.nexus-progress-bar-blue-label-span {
  color: var(--progress-bar-blue-label-span-color);
  font-size: var(--progress-bar-label-span-font-size);
}
