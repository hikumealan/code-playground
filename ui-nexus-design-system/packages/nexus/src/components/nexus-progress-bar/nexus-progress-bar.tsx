import { Component, h, Host, Prop, State, Watch } from '@stencil/core';

@Component({
  tag: 'nexus-progress-bar',
  shadow: false,
  styleUrl: 'nexus-progress-bar.scss'
})
export class NexusProgressBar {
  /* eslint-disable no-magic-numbers */
  private static readonly DEGREE_MULTIPLIER = 3.6;
  private static readonly PERCENT_MAX = 100;
  private static readonly PERCENT_INCREMENT_RATE = 15;
  /* eslint-enable no-magic-numbers */

  private interval: number;

  /**
   * Sets the current progress percentage out of 100 for the bar.
   */
  @Prop({ mutable: true }) value = 0;

  /**
   * Indicates if progress value should be ignored and the value indeterminate.
   */
  @Prop({ mutable: true }) indeterminate: boolean = false;

  /**
   * Indicates if the progress bar should present its error state.
   */
  @Prop({ mutable: true }) error: boolean = false;

  /**
   * Indicates if the progress indicator should be a circle instead of a bar.
   */
  @Prop({ mutable: true }) circle: boolean = false;

  @State() width: number;

  @State() hasError: boolean;

  @State() isIndeterminate: boolean;

  @State() isCircle: boolean;

  @Watch('error')
  setError(newValue: boolean, oldValue: boolean) {
    if (newValue === oldValue) {
      return;
    }

    this.hasError = newValue;
  }

  @Watch('indeterminate')
  setIndeterminate(newValue: boolean, oldValue: boolean) {
    if (newValue === oldValue) {
      return;
    }

    this.isIndeterminate = newValue;
    this.checkValue();
  }

  @Watch('value')
  setWidth(newValue: number, oldValue: number) {
    if (newValue === oldValue || this.isIndeterminate) {
      return;
    }

    this.width = newValue;
  }

  @Watch('circle')
  setCircle(newValue: boolean, oldValue: boolean) {
    if (newValue === oldValue) {
      return;
    }

    this.isCircle = newValue;
    this.checkValue();
  }

  checkValue() {
    if (this.isIndeterminate && this.isCircle) {
      this.updateValue();
    }
    else if (this.interval) {
      window.clearInterval(this.interval);
    }
  }

  updateValue() {
    this.interval = window.setInterval(() => {
      if (this.value === NexusProgressBar.PERCENT_MAX) {
        this.value = 0;
      }
      else {
        this.value++;
      }
    }, NexusProgressBar.PERCENT_INCREMENT_RATE);
  }

  componentWillLoad() {
    this.hasError = this.error;
    this.isIndeterminate = this.indeterminate;
    this.width = this.isIndeterminate ? 0 : this.value;
    this.isCircle = this.circle;
    this.checkValue();
  }

  progressBar() {
    return (
      <Host class="nexus-progress-bar">
        <slot />
        <div class="nexus-progress-bar-meter-container">
          <div
            class={{
              'nexus-progress-bar-meter': true,
              'nexus-progress-bar-meter-indeterminate': this.isIndeterminate,
              'nexus-progress-bar-meter-error': this.hasError
            }}
            style={{
              width: `${this.width}%`
            }}
          ></div>
        </div>
      </Host>
    );
  }

  progressCircle() {
    return (
      <Host class="nexus-progress-bar nexus-progress-bar-circle">
        <div
          class={{
            'nexus-progress-bar-pie': true,
            'nexus-progress-bar-hide-right':
              this.value <= NexusProgressBar.PERCENT_MAX / 2,
            'nexus-progress-bar-error': this.hasError
          }}
        >
          <div
            class="nexus-progress-bar-half-circle"
            style={{
              transform: `rotate(${this.value * NexusProgressBar.DEGREE_MULTIPLIER}deg)`
            }}
          />
          <div class={{
            'nexus-progress-bar-left-hidden': this.value <= NexusProgressBar.PERCENT_MAX / 2
          }}
          />
          <slot />
          <div class="nexus-progress-bar-right-side nexus-progress-bar-half-circle" />
        </div>

        <div class="nexus-progress-bar-shadow" />
      </Host>
    );
  }

  render() {
    return this.isCircle ? this.progressCircle() : this.progressBar();
  }
}
