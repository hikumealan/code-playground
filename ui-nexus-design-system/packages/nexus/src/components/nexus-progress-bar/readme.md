# nexus-progress-bar



<!-- Auto Generated Below -->


## Properties

| Property        | Attribute       | Description                                                                | Type      | Default |
| --------------- | --------------- | -------------------------------------------------------------------------- | --------- | ------- |
| `circle`        | `circle`        | Indicates if the progress indicator should be a circle instead of a bar.   | `boolean` | `false` |
| `error`         | `error`         | Indicates if the progress bar should present its error state.              | `boolean` | `false` |
| `indeterminate` | `indeterminate` | Indicates if progress value should be ignored and the value indeterminate. | `boolean` | `false` |
| `value`         | `value`         | Sets the current progress percentage out of 100 for the bar.               | `number`  | `0`     |


----------------------------------------------


