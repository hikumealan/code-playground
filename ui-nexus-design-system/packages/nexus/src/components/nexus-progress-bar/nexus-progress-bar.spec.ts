import { NexusProgressBar } from './nexus-progress-bar';
import { NexusProgressBarLabel } from './nexus-progress-bar-label/nexus-progress-bar-label';
import { newSpecPage } from '@stencil/core/testing';

const nexpbCls = 'nexus-progress-bar';
const nexpbmtrCls = '.nexus-progress-bar-meter';

describe('check nexus-progress-bar', () => {
  it('builds', async () => {
    const page = await newSpecPage({
      components: [NexusProgressBar],
      html: '<nexus-progress-bar></nexus-progress-bar>'
    });

    expect(new NexusProgressBar()).toBeTruthy();
    expect(page.root).toHaveClass(nexpbCls);
  });

  it('should set the name', async () => {
    const value = 50;

    const page = await newSpecPage({
      components: [NexusProgressBar],
      html: `<nexus-progress-bar value="${value}"></nexus-progress-bar>`
    });

    const progress = page.body.querySelector(nexpbCls);
    const meter = progress.querySelector(nexpbmtrCls);
    const style = meter.getAttribute('style');

    expect(progress.value).toEqual(value);
    expect(style.includes('50%')).toEqual(true);
  });

  it('should set the error state', async () => {
    const error = true;

    const page = await newSpecPage({
      components: [NexusProgressBar],
      html: `<nexus-progress-bar error="${error}"></nexus-progress-bar>`
    });
    const progress = page.body.querySelector(nexpbCls);
    const meter = progress.querySelector(nexpbmtrCls);

    expect(progress.error).toEqual(error);
    expect(meter).toHaveClass('nexus-progress-bar-meter-error');
  });

  it('should set the indeterminate state', async () => {
    const indeterminate = true;

    const page = await newSpecPage({
      components: [NexusProgressBar],
      html: `<nexus-progress-bar indeterminate="${indeterminate}"></nexus-progress-bar>`
    });
    const progress = page.body.querySelector(nexpbCls);
    const meter = progress.querySelector(nexpbmtrCls);

    expect(progress.indeterminate).toEqual(indeterminate);
    expect(meter).toHaveClass('nexus-progress-bar-meter-indeterminate');
  });

  it('should render with value', async () => {
    const value = 50;
    const page = await newSpecPage({
      components: [NexusProgressBar],
      html: `<nexus-progress-bar></nexus-progress-bar>`
    });

    page.rootInstance.value = value;
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('should render with error', async () => {
    const page = await newSpecPage({
      components: [NexusProgressBar],
      html: `<nexus-progress-bar></nexus-progress-bar>`
    });

    page.rootInstance.error = true;
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('should render with indeterminate', async () => {
    const page = await newSpecPage({
      components: [NexusProgressBar],
      html: `<nexus-progress-bar></nexus-progress-bar>`
    });

    page.rootInstance.indeterminate = true;
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('should have a label', async () => {
    const page = await newSpecPage({
      components: [NexusProgressBar,
        NexusProgressBarLabel],
      html: `<nexus-progress-bar value="50"><nexus-progress-bar-label>Label</nexus-progress-bar-label></nexus-progress-bar>`
    });
    const label = page.body.querySelector('nexus-progress-bar-label');
    expect(label).not.toBeNull();
  });

  it('should render with circle', async () => {
    const page = await newSpecPage({
      components: [NexusProgressBar],
      html: `<nexus-progress-bar></nexus-progress-bar>`
    });

    page.rootInstance.circle = true;
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('should render indeterminate circle', async (done) => {
    const page = await newSpecPage({
      components: [NexusProgressBar],
      html: `<nexus-progress-bar circle="true" indeterminate="true"></nexus-progress-bar>`
    });

    setTimeout(() => {
      // force clearing of value interval based on codepath
      page.rootInstance.indeterminate = false;
      // eslint-disable-next-line no-magic-numbers
    }, 1000);

    setTimeout(() => {
      page.waitForChanges();
      expect(page.root).toMatchSnapshot();
      done();
      // eslint-disable-next-line no-magic-numbers
    }, 1000);
  });
});
