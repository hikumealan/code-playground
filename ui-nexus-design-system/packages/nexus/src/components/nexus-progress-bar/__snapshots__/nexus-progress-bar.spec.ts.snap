// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check nexus-progress-bar should render indeterminate circle 1`] = `
<nexus-progress-bar circle="true" class="nexus-progress-bar nexus-progress-bar-circle" indeterminate="true">
  <!---->
  <div class="nexus-progress-bar-hide-right nexus-progress-bar-pie">
    <div class="nexus-progress-bar-half-circle" style="transform: rotate(0deg);"></div>
    <div class="nexus-progress-bar-left-hidden"></div>
    <div class="nexus-progress-bar-half-circle nexus-progress-bar-right-side"></div>
  </div>
  <div class="nexus-progress-bar-shadow"></div>
</nexus-progress-bar>
`;

exports[`check nexus-progress-bar should render with circle 1`] = `
<nexus-progress-bar class="nexus-progress-bar nexus-progress-bar-circle">
  <!---->
  <div class="nexus-progress-bar-hide-right nexus-progress-bar-pie">
    <div class="nexus-progress-bar-half-circle" style="transform: rotate(0deg);"></div>
    <div class="nexus-progress-bar-left-hidden"></div>
    <div class="nexus-progress-bar-half-circle nexus-progress-bar-right-side"></div>
  </div>
  <div class="nexus-progress-bar-shadow"></div>
</nexus-progress-bar>
`;

exports[`check nexus-progress-bar should render with error 1`] = `
<nexus-progress-bar class="nexus-progress-bar">
  <!---->
  <div class="nexus-progress-bar-meter-container">
    <div class="nexus-progress-bar-meter nexus-progress-bar-meter-error" style="width: 0%;"></div>
  </div>
</nexus-progress-bar>
`;

exports[`check nexus-progress-bar should render with indeterminate 1`] = `
<nexus-progress-bar class="nexus-progress-bar">
  <!---->
  <div class="nexus-progress-bar-meter-container">
    <div class="nexus-progress-bar-meter nexus-progress-bar-meter-indeterminate" style="width: 0%;"></div>
  </div>
</nexus-progress-bar>
`;

exports[`check nexus-progress-bar should render with value 1`] = `
<nexus-progress-bar class="nexus-progress-bar">
  <!---->
  <div class="nexus-progress-bar-meter-container">
    <div class="nexus-progress-bar-meter" style="width: 50%;"></div>
  </div>
</nexus-progress-bar>
`;